timemachine/fe/chiral_utils.py

```diff
@@ -55,3 +56,69 @@ def setup_chiral_bond_restraints(mol, conf, src_idx, dst_idx):
                 signs.append(-1)
 
     return np.array(restr_idxs), np.array(signs)
+
+
+def find_chiral_atoms(mol):
+    """
+    Find chiral atoms in a molecule. Note that an atom is chiral if it has an non-invertible
+    energy barrier. Even something like CH4 is considered chiral.
+
+    Parameters
+    ----------
+    mol: Chem.Mol
+        input molecule
+
+    Returns
+    -------
+    set of int
+        Chiral atoms
+
+    """
+    # these should be mutually exclusive, but if any pattern is hit then the results
+    # are accumulated to a set
+    chiral_patterns = [
+        "[X4:1]",  # any tetrahedral atom
+        "[#16X3,#15X3:1]",  # trivalent sulfur, phosphorous are assumed to be non-invertible
+        "[#7X3:1](~[R])(~[R])~[R]",  # nitrogen directly bonded to three ring atoms
+    ]
+
+    chiral_atoms = set()
+    for patt in chiral_patterns:
+        query_mol = Chem.MolFromSmarts(patt)
+        assert query_mol is not None
+        for match in mol.GetSubstructMatches(query_mol):
+            chiral_atoms.add(match[0])
+
+    return chiral_atoms
+
+
+def find_chiral_bonds(mol):
+    """
+    Find chiral bonds in a molecule:
+
+    Parameters
+    ----------
+    mol: Chem.Mol
+        input molecule
+
+    Returns
+    -------
+    set of 2-tuple
+        Chiral bonds
+
+    """
+
+    chiral_patterns = [
+        "[X2,X3:1]:[X2,X3:2]",  # any aromatic bond
```

Comment:
Specification: I don't understand why this pattern should be considered chiral. (I don't understand the goal that's achieved by saying there are 6 chiral bonds in benzene.) Deserves more elaboration!

Reply:
I was thinking about this from the geometric perspective, any planar bond group that has an extremely high rotational barrier (eg. amides). It seems like the aromatic ring bonds followed this rule of thumb as well. Unlike the chiral atom case, I guess the end-state for aromatic rings are probably well defined enough that even in the absence of chiral restraints they should adopt the correct geometry.

In any case, these restraints do not hurt (and are extremely fast to compute), and are similar to the butane case that it's probably fine to just leave them on?

Reply:
>I was thinking about this from the geometric perspective, any planar bond group that has an extremely high rotational barrier (eg. amides)

The motivation for amides makes perfect sense!

In general, if our MM energy function has multiple "wells", but we want to ensure we sample only one of them, this sort of restraint seems well-motivated.

(E.g. defining methane as "chiral" in your example -- hydrogen permutations would be iso-energetic, but (1) permutations are practically not sampled by MD for kinetic reasons, (2) we want to ensure we don't start sampling permutations unevenly if sampling becomes more thorough.)

>Unlike the chiral atom case, I guess the end-state for aromatic rings are probably well defined enough that even in the absence of chiral restraints they should adopt the correct geometry.

Yeah, for this reason the motivation for aromatic bonds is less obvious to me, since I don't see what the restraint is actually preventing. (H-C:C-H "inversions" in benzene would be extremely energetically unfavorable with or without the new restraints.)

>In any case, these restraints do not hurt (and are extremely fast to compute), and are similar to the butane case that it's probably fine to just leave them on?

Probably doesn't hurt! Would still be good to have a motivating example though, of some "inversion" that's energetically allowed in the absence of the aromatic bond restraint, that the restraint now rules out.

Reply:
For the sake of simplicity - I'll remove this for now. We can add this later on if needed.